# String and Number
In this notebook we will study Python hello world, then the numbers and strings.

# Prerequisites

- Assume you have a Windows laptop with internet connection.
- You need install Visual Studio Code from Microsoft Store.
- Open a folder, such as Pyhton in your machine and create a notebooks folder under it.
- Copy any hello world ipynb file in the notebooks folder.
- Launch VSCode and open this folder.
- Install latest Python + Jupyter notebook when click install kernel.
- Set default environment from Get Started with Python Development.

## Hello World
The following python code will first print a line of text "Hello World"
Then we will assign this sentence to a variable and print it again.

In [1]:
# Print Hello world. 
print('Hello World!')

# Assign it to a variable and print it again.
greeting_message = "Hello World!"
print(greeting_message)

Hello World!
Hello World!


#### Observation
From the above example, we observe following points:
1. Python does not use any statement end mark (for example ; in C/C++), end of line means end of statement.
2. We do not need to declare the type of a variable in advance (but we can do it, we will discuss later).

#### Variable Naming Convention 
We use lower case as name of variable, if the variable contains two english words, we use underscore '_' to connect them.

## String
Now let's look at string, a line of text.

In [2]:
# Let's first get a variable
name = "ada lovelace"
print(name)

# Make upper case for the first character in each word.
print(name.title())

# Make it all upper case
print(name.upper())
print(name.lower())

# Let's use variable as placeholder in a string
greeting_message = 'Hi, my name is {name_placeholder}'

# replace the placeholder by real name
greeting_message = greeting_message.format(name_placeholder=name.title())
print(greeting_message)

ada lovelace
Ada Lovelace
ADA LOVELACE
ada lovelace
Hi, my name is Ada Lovelace


#### Observation 
All the string function above will not change the input string itself, instead it will return the result of operation.

#### Space
Now Let's look at same common space characters in the string. Then try to remove the extra space before and after the text. 

In [3]:
# \t is the tab key and \n is the new line
name = "My name is\nada \t lovelace"
print(name)

# Remove the space before and after.
string = '   \t python  '
# we see spaces before and after
print(string)
# we remove spaces before the string
print(string.lstrip())
# we remove spaces after the string
print(string.rstrip())
# we remove spaces before and after the string
print(string.strip())

My name is
ada 	 lovelace
   	 python  
python  
   	 python
python


### Quotes
In python we can use either double quotes or single quotes. In this case you can avoid escape the special characters.
We can also specify a multi-lines text to a string.


In [4]:
# Let's use single quote inside a string
sentence = "One of Python's strength is its diverse community."
print(sentence)

# Let's create a multi-line text with any special characters inside it.
text = """
The Zen of Python, by Tim Peters:
"Beautiful is better than ugly. 
Simple is better than complex.
Complex is better than complicated.
Readability counts
"
"""
print(text)

One of Python's strength is its diverse community.

The Zen of Python, by Tim Peters:
"Beautiful is better than ugly. 
Simple is better than complex.
Complex is better than complicated.
Readability counts
"



## Numbers
Let's start some numeric value in python. 

In [7]:
# Let's work on some basic arithmetic
a = 2 + 3
print(a)

a = 3 - 2
print(a)

a = 2 * 3
print(a)

a = 5 / 2
print('5 / 2 = ' + str(a))
print('int(5 / 2) = ' + str(int(a)))

a = 3**2
print('3**2' + str(a))

a = 2 * (4-2)
print(a)

a = 0.1 + 0.1
print(a)

a = 0.2 + 0.1
print(a)

large = 14_000_000_000_000_000_000_000_000_000_000_000
print(large)

x, y, z = 1, 2, 3
print(x,y,z)

# The following is a constant value
MAX_CONNECTIONS = 5000
print('MAX_CONNECTIONS' + str(MAX_CONNECTIONS))

# calculate module
print('7 % 3 = ' + str(7%3))


5
1
6
5 / 2 = 2.5
int(5 / 2) = 2
3**29
4
0.2
0.30000000000000004
14000000000000000000000000000000000
1 2 3
MAX_CONNECTIONS5000
7 % 3 = 1


#### Observation 
1. In Python there is no difference between integer and float. By default, the division result will be treated as float. This is quite different from other language, for example in C/C++ or Java, 5/2 = 2, not 2.5, but in python it is 2.5
2. Like all other languages, float will encounter precision issue, for example like 0.1 + 0.2. We will study precision round up later.
3. Python can use under score '_' to separate multiple digits in a number.
4. Python does not have capacity for a number, unlike C/C++ or Java, which will have integer(32 bits) and long (64 bits). A numeric value in python can be very big. But it is also not a double value, so you should not worry about precision of the large number.
5. We can assign multiple values to multiple variables in one statement.
6. Constant value can be represented by all upper case variables.
7. We can convert a float to integer by int().
8. We can convert numeric to string by str()